### PR TITLE
Add unit test for frontend/admin/handler - part 2

### PR DESCRIPTION
**Pull request by neil-xie**
**What changed?**
Add more unit test for frontend/admin/handler

**Why?**

**How did you test it?**

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: service/frontend/admin/handler_test.go [hunk 1]
```diff
@@ -1619,3 +1621,950 @@ func TestGetReplicationMessages(t *testing.T) {
...
+				ClusterName:            "test-cluster",
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+			},
+			drqHandlerFunc: func(mock *domain.MockReplicationQueue) {
+				mock.EXPECT().GetAckLevels(gomock.Any()).Return(map[string]int64{
+					"test-cluster": 1,
+				}, nil)
+				mock.EXPECT().GetReplicationMessages(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, int64(2), nil)
+				mock.EXPECT().UpdateAckLevel(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil)
```
Comment (davidporter-id-au) [inline service/frontend/admin/handler_test.go:1650]: I would suggest not using goock.Any() on write operations like this where you want to ensure that if you were to do a big refactor after this test was landed, that if they were to get zero'd out or were wrong or something that the test would still pass. 

Asserting - particularly in write cases - at least the important values, at both ends is probably worthwhile

File: service/frontend/admin/handler_test.go [hunk 2]
```diff
@@ -1619,3 +1621,950 @@ func TestGetReplicationMessages(t *testing.T) {
...
+		"normal request": {
+			input: &types.ReapplyEventsRequest{
+				DomainName: "test-domain",
+				WorkflowExecution: &types.WorkflowExecution{
+					WorkflowID: "test-workflow-id",
+				},
+				Events: &types.DataBlob{},
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().ReapplyEvents(gomock.Any(), gomock.Any()).Return(nil)
```
Comment (davidporter-id-au) [inline service/frontend/admin/handler_test.go:1849]: same as above, you may want to be a little more specific about what's accepted

File: service/frontend/admin/handler_test.go [hunk 3]
```diff
@@ -1619,3 +1621,950 @@ func TestGetReplicationMessages(t *testing.T) {
...
+			if td.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_SerializeRawHistoryToken(t *testing.T) {
```
Comment (davidporter-id-au) [inline service/frontend/admin/handler_test.go:2505]: consider doing a round-trip test on serializations. It is usually easier too. If you can serialize a struct, then unserialise it and then assert that they're equal you'll catch a lot of bugs. Look at `common/testdatagen` if you want an easy way to make data for round-trip tests

Comment (neil-xie) [reply service/frontend/admin/handler_test.go:2505]: Right, it is much better than testing separately

File: service/frontend/admin/handler_test.go [hunk 4]
```diff
@@ -1619,3 +1621,1120 @@ func TestGetReplicationMessages(t *testing.T) {
...
+						FirstEventID: 1,
+					},
+				},
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().GetDLQReplicationMessages(gomock.Any(), gomock.Any()).Return(nil, nil)
+			},
+			wantErr: false,
+		},
+		"get dql message error": {
```
Comment (taylanisikdemir) [inline service/frontend/admin/handler_test.go:1749]: nit typo: dlq

File: service/frontend/admin/handler_test.go [hunk 5]
```diff
@@ -1619,3 +1621,1120 @@ func TestGetReplicationMessages(t *testing.T) {
...
+						RunID:        uuid.New(),
+						TaskID:       int64(1),
+						TaskType:     1,
+						Version:      0,
+						FirstEventID: 1,
+					},
+				},
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().GetDLQReplicationMessages(gomock.Any(), gomock.Any()).Return(nil, nil)
```
Comment (taylanisikdemir) [inline service/frontend/admin/handler_test.go:1745]: let's return some non-nil object here and validate it's received in test code

File: service/frontend/admin/handler_test.go [hunk 6]
```diff
@@ -1619,3 +1621,1120 @@ func TestGetReplicationMessages(t *testing.T) {
...
+				handler.throttleRetry = backoff.NewThrottleRetry(
+					backoff.WithRetryPolicy(policy),
+					backoff.WithRetryableError(func(_ error) bool { return true }),
+				)
+				mockDlqHandler := domain.NewMockDLQMessageHandler(goMock)
+				mockDlqHandler.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil, nil)
+				handler.domainDLQHandler = mockDlqHandler
+			}
+
+			_, err := handler.ReadDLQMessages(context.Background(), td.input)
```
Comment (taylanisikdemir) [inline service/frontend/admin/handler_test.go:2000]: same here. missing validation of returned messages for success case

File: service/frontend/admin/handler_test.go [hunk 7]
```diff
@@ -1619,3 +1621,1120 @@ func TestGetReplicationMessages(t *testing.T) {
...
+			handler := adminHandlerImpl{
+				Resource: &resource.Test{
+					Logger:        testlogger.New(t),
+					MetricsClient: metrics.NewNoopMetricsClient(),
+					HistoryClient: hcMock,
+				},
+				domainDLQHandler: dlqMock,
+			}
+
+			_, err := handler.CountDLQMessages(context.Background(), td.input)
```
Comment (taylanisikdemir) [inline service/frontend/admin/handler_test.go:2156]: same here

File: service/frontend/admin/handler_test.go [hunk 8]
```diff
@@ -1619,3 +1621,1120 @@ func TestGetReplicationMessages(t *testing.T) {
...
+			if td.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_GetCrossClusterTasks(t *testing.T) {
```
Comment (taylanisikdemir) [inline service/frontend/admin/handler_test.go:2398]: I think @davidporter-id-au will cleanup cross cluster related code but doesn't hurt to keep these new tests until that happens

**Review by taylanisikdemir — APPROVED** [review]
LGTM with comments

**Pull request merged**